python/_tskitmodule.c

```diff
@@ -9860,6 +9860,110 @@ TreeSequence_divergence_matrix(TreeSequence *self, PyObject *args, PyObject *kwd
     return ret;
 }
 
+static PyObject *
+TreeSequence_ld_matrix(TreeSequence *self, PyObject *args, PyObject *kwds,
+    two_locus_count_stat_method *method)
+{
+    PyObject *ret = NULL;
+    static char *kwlist[]
+        = { "sample_set_sizes", "sample_sets", "row_sites", "col_sites", "mode", NULL };
+
+    PyObject *sample_set_sizes = NULL;
+    PyObject *sample_sets = NULL;
+    PyArrayObject *sample_set_sizes_array = NULL;
+    PyArrayObject *sample_sets_array = NULL;
+    PyArrayObject *row_sites = NULL;
+    PyArrayObject *col_sites = NULL;
+    PyArrayObject *result_matrix = NULL;
+    npy_intp result_shape[3];
+    char *mode = NULL;
+    tsk_size_t num_sample_sets;
+    tsk_flags_t options = 0;
+    int err;
+
+    if (TreeSequence_check_state(self) != 0) {
+        goto out;
+    }
+    if (!PyArg_ParseTupleAndKeywords(args, kwds, "OOOOs", kwlist, &sample_set_sizes,
+            &sample_sets, &row_sites, &col_sites, &mode)) {
+        goto out;
+    }
+    if (parse_stats_mode(mode, &options) != 0) {
+        goto out;
+    }
+    if (parse_sample_sets(sample_set_sizes, &sample_set_sizes_array, sample_sets,
+            &sample_sets_array, &num_sample_sets)
+        != 0) {
+        goto out;
+    }
+
+    result_shape[0] = PyArray_DIM(row_sites, 0);
```

Comment:
You're right, it wasn't safe. I was controlling the input data in the outer python layer, but it makes sense to prevent segfaulting at this layer too. I've added an `asarray` call, which helpfully makes assertions about casting.